sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextRowCountEstimator.java

```diff
@@ -0,0 +1,154 @@
+  }
+
+  /** Builder for {@link org.apache.beam.sdk.io.TextRowCountEstimator}. */
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder setNumSampledFiles(int numSampledFiles);
+
+    public abstract Builder setNumSampledBytes(long numSampledBytes);
+
+    public abstract Builder setDirectoryTreatment(
+        FileIO.ReadMatches.DirectoryTreatment directoryTreatment);
+
+    public abstract Builder setCompression(Compression compression);
+
+    public abstract Builder setDelimiters(byte[] delimiters);
+
+    public abstract Builder setFilePattern(String filePattern);
+
+    public abstract Builder setEmptyMatchTreatment(EmptyMatchTreatment emptyMatchTreatment);
+
+    public Builder sampleAllFiles() {
+      return this.setNumSampledFiles(-1);
+    }
+
+    public abstract TextRowCountEstimator build();
+  }
+
+  /**
+   * Estimates the number of non empty rows. Note that it does not cache the previous estimate. It
+   * returns 0L if all the lines that have been read are empty.
+   *
+   * @return Number of estimated rows.
+   * @throws IOException
+   */
+  public Long estimateRowCount(PipelineOptions pipelineOptions) throws IOException {
+    long linesSize = 0;
+    int numberOfReadLines = 0;
+    long totalFileSizes = 0;
+    int numberOfReadFiles = 0;
+
+    MatchResult match = FileSystems.match(getFilePattern(), getEmptyMatchTreatment());
+
+    for (MatchResult.Metadata metadata : match.metadata()) {
+
+      if ((!isSamplingAllFiles()) && numberOfReadFiles > getNumSampledFiles()) {
+        break;
+      }
+
+      FileIO.ReadableFile file =
+          FileIO.ReadMatches.matchToReadableFile(
+              metadata, getDirectoryTreatment(), getCompression());
+
+      long readingWindowSize = Math.min(getNumSampledBytes(), metadata.sizeBytes());
+      OffsetRange range = new OffsetRange(0, readingWindowSize);
+
+      TextSource textSource =
+          new TextSource(
+              ValueProvider.StaticValueProvider.of(file.getMetadata().resourceId().toString()),
+              getEmptyMatchTreatment(),
+              getDelimiters());
+      FileBasedSource<String> source =
+          CompressedSource.from(textSource).withCompression(file.getCompression());
+      try (BoundedSource.BoundedReader<String> reader =
+          source
+              .createForSubrangeOfFile(file.getMetadata(), range.getFrom(), range.getTo())
+              .createReader(pipelineOptions)) {
+        for (boolean more = reader.start(); more; more = reader.advance()) {
+          int length = reader.getCurrent().length();
+          linesSize += reader.getCurrent().length() + 1; // for \n at the end of line
+          numberOfReadLines += length == 0 ? 0 : 1;
+        }
+      }
+      long fileSize = metadata.sizeBytes();
+      numberOfReadFiles += fileSize == 0 ? 0 : 1;
+      totalFileSizes += fileSize;
+    }
+
+    if (totalFileSizes == 0) {
+      return 0L;
+    }
+
+    if (numberOfReadLines == 0) {
```

Comment:
Is it safe to only check `numberOfReadLines`? If it's zero, then why do we care about number of files?

Reply:
I think we can remove the previous if clause. Previously I was returning null in case of numberOfReadLines==0, and I forgot to fix this part after changing that.
I was thinking about returning null when we are not sure about the estimate (we haven't read non empty lines) and return 0 when we are sure that the number of lines is 0. I will change this part and push another commit.